x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/ingest/CircleProcessor.java

```diff
@@ -0,0 +1,198 @@
+            valueWrapper = Map.of("shape", fieldValue);
+        } else if (obj instanceof String) {
+            valueWrapper = Map.of("shape", obj);
+        } else {
+            throw new IllegalArgumentException("field [" + field + "] must be a WKT Circle or a GeoJSON Circle value");
+        }
+
+        MapXContentParser parser = new MapXContentParser(NamedXContentRegistry.EMPTY,
+            DeprecationHandler.THROW_UNSUPPORTED_OPERATION, valueWrapper, XContentType.JSON);
+        try {
+            parser.nextToken(); // START_OBJECT
+            parser.nextToken(); // "shape" field key
+            parser.nextToken(); // shape value
+            GeometryFormat geometryFormat = PARSER.geometryFormat(parser);
+            Geometry geometry = geometryFormat.fromXContent(parser);
+            if (ShapeType.CIRCLE.equals(geometry.type())) {
+                Circle circle = (Circle) geometry;
+                int numSides = numSides(circle.getRadiusMeters());
+                Geometry polygonizedCircle = createRegularPolygon(circle.getLat(), circle.getLon(), circle.getRadiusMeters(), numSides);
+                XContentBuilder newValueBuilder = XContentFactory.jsonBuilder().startObject().field("val");
+                geometryFormat.toXContent(polygonizedCircle, newValueBuilder, ToXContent.EMPTY_PARAMS);
+                newValueBuilder.endObject();
+                Map<String, Object> newObj = XContentHelper.convertToMap(
+                    BytesReference.bytes(newValueBuilder), true, XContentType.JSON).v2();
+                ingestDocument.setFieldValue(targetField, newObj.get("val"));
+            } else {
+                throw new IllegalArgumentException("found [" + geometry.type() + "] instead of circle");
+            }
+        } catch (Exception e) {
+            throw new IllegalArgumentException("invalid circle definition", e);
+        }
+
+        return ingestDocument;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    String field() {
+        return field;
+    }
+
+    String targetField() {
+        return targetField;
+    }
+
+    double errorDistanceMeters() {
+        return errorDistanceMeters;
+    }
+
+    int numSides(double radiusMeters) {
+        int val = (int) Math.ceil(2 * Math.PI / Math.acos(1 - errorDistanceMeters / radiusMeters));
+        return Math.min(MAXIMUM_NUMBER_OF_SIDES, Math.max(MINIMUM_NUMBER_OF_SIDES, val));
+    }
+
+
+    /**
+     * Makes an n-gon, centered at the provided lat/lon, and each vertex approximately
+     * radiusMeters away from the center.
+     *
+     * This does not split the polygon across the date-line. Relies on {@link GeoShapeIndexer} to
+     * split prepare polygon for indexing.
+     *
+     * Adapted from from org.apache.lucene.geo.GeoTestUtil
+     * */
+    public static Polygon createRegularPolygon(double centerLat, double centerLon, double radiusMeters, int gons) {
```

Comment:
Would it make sense to move this method to some SpatialUtils utility class? I feel like it's pretty generic and we might find some other ways to use it. I think I would also replace first three doubles with Circle. And we should figure out what to do with the radiusMeters parameter in Circle since it is not meters in case of `shape`, but this is a topic for another PR.

Reply:
I've moved things to SpatialUtils 👍